src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -46,26 +48,53 @@ internal sealed partial class LocalRewriter
             }
         }
 
-        private BoundExpression VisitArrayOrSpanCollectionExpression(BoundCollectionExpression node, TypeSymbol collectionType, TypeWithAnnotations elementType)
+        private BoundExpression VisitArrayOrSpanCollectionExpression(BoundCollectionExpression node, CollectionExpressionTypeKind collectionTypeKind, TypeSymbol collectionType, TypeWithAnnotations elementType)
         {
             Debug.Assert(!_inExpressionLambda);
+            Debug.Assert(_additionalLocals is { });
 
             var syntax = node.Syntax;
+            var elements = node.Elements;
             MethodSymbol? spanConstructor = null;
 
             var arrayType = collectionType as ArrayTypeSymbol;
             if (arrayType is null)
             {
-                Debug.Assert(collectionType.Name is "Span" or "ReadOnlySpan");
                 // We're constructing a Span<T> or ReadOnlySpan<T> rather than T[].
                 var spanType = (NamedTypeSymbol)collectionType;
+
+                Debug.Assert(collectionTypeKind is CollectionExpressionTypeKind.Span or CollectionExpressionTypeKind.ReadOnlySpan);
+                Debug.Assert(spanType.OriginalDefinition.Equals(_compilation.GetWellKnownType(
+                    collectionTypeKind == CollectionExpressionTypeKind.Span ? WellKnownType.System_Span_T : WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.AllIgnoreOptions));
                 Debug.Assert(elementType.Equals(spanType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0], TypeCompareKind.AllIgnoreOptions));
+
+                if (collectionTypeKind == CollectionExpressionTypeKind.ReadOnlySpan &&
+                    ShouldUseRuntimeHelpersCreateSpan(node, elementType.Type))
```

Comment:
I'm not sure what all the requirements to be able to emit as a blob are. Is there any case where emitting as a blob could fail (thus fall back to a plain array instantiation) but we could emit as an inline array below?

Reply:
If the collection is a candidate to be emitted in the assembly data section - that is, if it is a `ReadOnlySpan<T>` where `T` is one of the expected primitive types, and the collection contains constant values only - then it will be emitted as global data. The only condition that should affect the generated code is whether the particular `RuntimeHelpers.CreateSpan()` optimization is available, but the result is still global data. See `CollectionExpressionTests.RuntimeHelpers_CreateSpan` for instance.